### DISCUSSION TITLE
Adding Polish characters to a French keyboard layout

### DISCUSSION THREAD

**Discussion by lanigb** [fcitx/fcitx5#1008] [Q&A] [answered]
Hi all,

I use as a default the fr-oss layout (azerty). I would like to replace unused Alt-gr combinations for Polish only characters (there are few of them). How can I do that ?
I can't use the Polish input methods because they are mapped to a qwerty layout.

Thanks for your help.

**Comment by wengxt** [comment]
![图片](https://github.com/fcitx/fcitx5/assets/259684/a1e4d461-9bd3-4599-8fb1-1bef16cae229)

fr-oss layout looks like this. I don't really see there's any "unused" ones.

Are you using X11, or wayland?

**Reply by lanigb** [reply]
Thanks for your answer.
I should have written « unused in French » or « that I don't use » .
I don't write Spanish nor German or Latin so I don't use ß, æ, ¿ and some others.  Especially, any vowel  with ¨ can be written with the compose key (that what I usually use)  so they are good candidates for me for Polish letters.

**Reply by wengxt** [reply]
Have you considered using Compose for polish characters? Compose sequence can be easily overridden at user level.

For example if you don’t need key ð
you can simply write 
```
include "%L"
<eth> : "ABCD"
```
 Then AltGr + H on fr-oss will produce ABCD

Creating a new xkb layout requires more work to plumb all the required files. Especially Xserver doesn’t really read user level layout files.

xmodmap could work and fcitx does support to load it for you, but it’s more fragile and not supported by wayland.

Or fcitx also support a long press option that’s very customizable, but might feels harder to use (you have to long press and then select) if you need to type those character frequently.

**Reply by lanigb** [reply]
This is the first thing I tried but with no success.
I made a minimal  .XCompose
```
# Define AltGr + a to produce Ą
<Multi_key> <a> : "Ą" Aogonek
```
Nothing changed.  Is my syntax wrong ?

**Reply by wengxt** [reply]
I'm not sure why you put <Multi_key> there.
First of all, "<...>" is to describe a key sym name. <Multi_key> is NOT a special syntax, it's just a key that really produce <Multi_key>.

If you didn't tune any xkb option, there's no key on fr-oss will produce <Multi_key>.

You can start xev and press the key and see the key sym name.

In your case, press altgr+a should show

```
KeyPress event, serial 39, synthetic NO, window 0x1600001,
    root 0x3cf, subw 0x0, time 121658985, (-201,199), root:(669,642),
    state 0x80, keycode 24 (keysym 0xe6, ae), same_screen YES,
    XLookupString gives 2 bytes: (c3 a6) "æ"
    XmbLookupString gives 2 bytes: (c3 a6) "æ"
    XFilterEvent returns: False
```

Then I will put
```
<ae> : "Ą" 
```

If you want to put comment,  "Aogonek" you forget "#"
```
<ae> : "Ą"  # Aogonek
```

Also, the line `include "%L"` will ensure you include the system built-in Compose sequence (in most locale, it will be `/usr/share/X11/locale/en_US.UTF-8/Compose` )

**Reply by wengxt** [reply]
Video demo on how this works.

https://github.com/fcitx/fcitx5/assets/259684/2b6afa9d-7a26-4264-a9ae-9689fa841ec3

**Reply by wengxt** [reply]
One last reminder, for now, you have to restart fcitx5 to make it reload ~/.XCompose file.

**Reply by lanigb** [reply]
I have got the Ą working !
Many thanks !

**Reply by lanigb** [reply, marked as answer]
In case it could be useful to someone, I post my .XCompose
```
  7 include "%L"                                                                             
  6 <ae> : "ą"                                                                               
  5 <AE> : "Ą"                                                                               
  4 <copyright> : "ć"                                                                        
  3 <registered> : "Ć"                                                                       
  2 <acircumflex> : "ź"                                                                      
  1 <Acircumflex> : "Ź"                                                                      
  0 <U0140> : "ł"                                                                            
  1 <U013F> : "Ł"                                                                            
  2 <notsign> : "ń"                                                                          
  3 <rightarrow> : "Ń"                                                                       
  4 <ocircumflex> : "ó"#P mappé sur P car                                                    
  5 <Ocircumflex> : "Ó" # alt-gr O = œ                                                       
  6 <ssharp> : "ś"                                                                           
  7 <doublelowquotemark> : "Ś"                                                               
  8 <ecircumflex> : "ę" #R mappé sur la touche voisine                                       
  9 <Ecircumflex> : "Ę"# de e car alt-gr e = €                                               
 10 <thorn> : "ż" # T car Z déjà pris                                                        
 11 <THORN> : "Ż"    
```
Wengxt, Thanks again for your kind help.

**Comment by directentis1** [comment]
@lanigb Hi, can I ask you a side question? How can I select the fr-oss layout in fcitx5? I don't see this option in my Ubuntu 22.04.

**Reply by lanigb** [reply]
It is called "clavier Français-Français (variante). But, in Debian,  at install time I have chosen this keyboard as being the main one.

**Reply by directentis1** [reply]
Got it, in the English user interface, it's “Keyboard - French - French (alt.),” even though it doesn't appear as fr-oss in the tooltip when switching input methods.

<img width="800" height="600" alt="image" src="https://github.com/user-attachments/assets/d8e7df4e-20db-4dba-abe3-8cb056db2a28" />
